base_location/models/better_zip.py

```diff
@@ -11,37 +12,108 @@ class BetterZip(models.Model):
     _name = "res.better.zip"
     _description = __doc__
     _order = "name asc"
-    _rec_name = "display_name"
 
-    display_name = fields.Char('Name', compute='_get_display_name', store=True)
+    def domain_city_id(self):
+        if self.country_id:
+            return [('country_id', '=', self.country_id.id)]
+        return ""
+
+    def domain_state_id(self):
+        if self.country_id:
+            return [('country_id', '=', self.country_id.id)]
+        return ""
+
     name = fields.Char('ZIP')
-    code = fields.Char('City Code', size=64,
-                       help="The official code for the city")
+    code = fields.Char(
+        'City Code',
+        size=64,
+        help="The official code for the city"
+    )
     city = fields.Char('City', required=True)
-    state_id = fields.Many2one('res.country.state', 'State')
+    city_id = fields.Many2one(
+        'res.city',
+        'City',
+    )
+    state_id = fields.Many2one(
+        'res.country.state',
+        'State',
+    )
     country_id = fields.Many2one('res.country', 'Country')
+    enforce_cities = fields.Boolean(
+        related='country_id.enforce_cities',
+        readonly=True,
+    )
     latitude = fields.Float()
     longitude = fields.Float()
 
-    @api.one
-    @api.depends(
-        'name',
-        'city',
-        'state_id',
-        'country_id',
-        )
-    def _get_display_name(self):
-        if self.name:
-            name = [self.name, self.city]
-        else:
-            name = [self.city]
-        if self.state_id:
-            name.append(self.state_id.name)
+    @api.multi
+    @api.depends('name', 'city', 'state_id', 'country_id')
+    def name_get(self):
+        result = []
+        for rec in self:
+            name = []
+            if rec.name:
+                name.append(rec.name)
+            name.append(rec.city)
+            if rec.state_id:
+                name.append(rec.state_id.name)
+            if rec.country_id:
+                name.append(rec.country_id.name)
+            result.append((rec.id, ", ".join(name)))
```

Comment:
Shouldn't you be using a `res.lang`-specific template if available? Such as `%(city)s, %(state)s (%(country)s)` for example